### PR TITLE
Flutter & Dart SDK refactor

### PR DESCRIPTION
**Pull request by lohanidamodar**
Refactoring Flutter SDK and Introduce Realtime Service

1. Refactor, remove `dio` dependency, instead use `http`
2. Client is refactored for proper cross-platform support
3. Adds a `Realtime` service that exposes `subscribe` method and allows to subscribe to Appwrite's events in realtime
4. Realtime service too, internally is built for proper cross-platform support

### CODE REVIEW

File: templates/dart/lib/src/client.dart.twig [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import 'response.dart';
+
+abstract class Client {
+  late Map<String, String> config;
+  late String _endPoint;
+
+  String get endPoint => _endPoint;
+
+  factory Client(
+          {String endPoint = 'https://appwrite.io/v1',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client.dart.twig:14]:
```suggestion
          {String endPoint = '{{ spec.endpoint }}',
```

File: templates/dart/lib/src/client.dart.twig [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  Client setSelfSigned({bool status = true});
+
+  Client setEndpoint(String endPoint);
+
+{% for header in spec.global.headers %}
+{% if header.description %}
+    /// {{header.description}}
+{% endif %}
+  Client set{{header.key | caseUcfirst}}(value);
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client.dart.twig:26]:
```suggestion
  /// {{header.description}}
{% endif %}
  Client set{{header.key | caseUcfirst}}(value);
```

to align the comment 👍🏻

File: templates/dart/lib/src/client_base.dart.twig [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import 'response.dart';
+import 'client.dart';
+import 'enums.dart';
+
+abstract class ClientBase implements Client {  
+{% for header in spec.global.headers %}
+{% if header.description %}
+    /// {{header.description}}
+{% endif %}
+  ClientBase set{{header.key | caseUcfirst}}(value);
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_base.dart.twig:10]: Or is it necessary that the comment has more indentation?

File: templates/dart/lib/src/client_browser.dart.twig [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    ClientBrowser(endPoint: endPoint, selfSigned: selfSigned);
+
+class ClientBrowser extends ClientBase with ClientMixin {
+  String _endPoint;
+  Map<String, String>? _headers;
+  late Map<String, String> config;
+  late BrowserClient _httpClient;
+
+  ClientBrowser({
+    String endPoint = 'https://appwrite.io/v1',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_browser.dart.twig:23]:
```suggestion
    String endPoint = '{{ spec.endpoint }}',
```

File: templates/dart/lib/src/client_browser.dart.twig [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    bool selfSigned = false,
+  }) : _endPoint = endPoint {
+    _httpClient = BrowserClient();
+    _headers = {
+      'content-type': 'application/json',
+      'x-sdk-version': 'appwrite:flutter:0.6.4',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_browser.dart.twig:29]:
```suggestion
      'x-sdk-version': '{{spec.title | caseDash}}:{{ language.name | caseLower }}:{{ sdk.version }}',
```

File: templates/dart/lib/src/client_browser.dart.twig [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      'X-Appwrite-Response-Format': '0.8.0',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_browser.dart.twig:30]: Here you can use 

```twig
{% for key,header in spec.global.defaultHeaders %}
        '{{key}}': '{{header}}',
{% endfor %}
```

to get the global headers 👍🏻

File: templates/dart/lib/src/client_io.dart.twig [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+class ClientIO extends ClientBase with ClientMixin {
+  String _endPoint;
+  Map<String, String>? _headers;
+  late Map<String, String> config;
+  bool _initialized = false;
+  late http.Client _httpClient;
+  late HttpClient _nativeClient;
+
+  ClientIO({
+    String endPoint = 'https://appwrite.io/v1',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_io.dart.twig:28]:
```suggestion
    String endPoint = '{{ spec.endpoint }}',
```

File: templates/dart/lib/src/client_io.dart.twig [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  }) : _endPoint = endPoint {
+    _nativeClient = new HttpClient()
+      ..badCertificateCallback =
+          ((X509Certificate cert, String host, int port) => selfSigned);
+    _httpClient = new IOClient(_nativeClient);
+    this._endPoint = endPoint;
+    this._headers = {
+      'content-type': 'application/json',
+      'x-sdk-version': 'appwrite:flutter:0.6.4',
+      'X-Appwrite-Response-Format': '0.8.0',
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/client_io.dart.twig:39]: Same changes that I reviewed in `client_browser.dart.twig` 👍🏻

File: templates/dart/lib/src/exception.dart.twig [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
   final String? message;
   final int? code;
   final dynamic response;
 
   {{spec.title | caseUcfirst}}Exception([this.message = "", this.code, this.response]);
   
-}
\ No newline at end of file
+  String toString() {
+    if (message == null) return "AppwriteException";
```
Comment (TorstenDittmann) [inline templates/dart/lib/src/exception.dart.twig:9]: Let's avoid hardcoded `Appwrite` :-)

File: templates/flutter/lib/src/client.dart.twig [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import 'response.dart';
+
+abstract class Client {
+  late Map<String, String> config;
+  late String _endPoint;
+
+  String get endPoint => _endPoint;
+
+  factory Client(
+          {String endPoint = 'https://appwrite.io/v1',
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/client.dart.twig:14]: See the previous comments 👍🏻

**Review by TorstenDittmann — CHANGES_REQUESTED** [review]
Awesome work so far @lohanidamodar 👏🏻

File: templates/dart/lib/src/client_base.dart.twig (hunk 3 above)
Comment (lohanidamodar) [reply templates/dart/lib/src/client_base.dart.twig:10]: Nope

File: templates/flutter/lib/src/realtime_io.dart.twig [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        ? await _connectForSelfSignedCert(uri, headers)
+        : await WebSocket.connect(uri.toString(), headers: headers));
+    return _websok;
+  }
+
+  /// Subscribe
+  ///
+  /// Use this method to subscribe to a channels and listen to
+  /// realtime events on those channels
+  RealtimeSubscription subscribe(List<String> channels) {
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/realtime_io.dart.twig:41]: We also need to allow passing a single string, that we will convert to an array 👍🏻

File: templates/flutter/lib/src/realtime_mixin.dart.twig [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      if (!this._channels.containsKey(channel)) {
+        this._channels[channel] = [];
+      }
+      this._channels[channel]!.add(controller);
+    });
+    Future.delayed(Duration.zero, () => _createSocket());
+    RealtimeSubscription subscription = RealtimeSubscription(
+        stream: controller.stream,
+        close: () {
+          controller.close();
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/realtime_mixin.dart.twig:92]: Is it possible to have the same method signature as Web and Kotlin?

Meaning `subscribe(channel, callback)`

Comment (lohanidamodar) [reply templates/flutter/lib/src/realtime_mixin.dart.twig:92]: But, stream is the proper way to listen to live changes in Flutter/Dart, I don't think using callback would give proper experience.

File: templates/flutter/lib/src/realtime_io.dart.twig (hunk 11 above)
Comment (lohanidamodar) [reply templates/flutter/lib/src/realtime_io.dart.twig:41]: @TorstenDittmann Dart doesn't allow multiple definitions of same function, will need a separate name is that a good idea? if yes, what would be the good name ?

File: templates/flutter/lib/src/client.dart.twig [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+  Client setSelfSigned({bool status = true});
+
+  Client setEndpoint(String endPoint);
+
+  Client setEndPointRealtime(String endPoint);
+
+{% for header in spec.global.headers %}
+{% if header.description %}
+    /// {{header.description}}
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/client.dart.twig:30]: Indentation 👍🏻

File: templates/flutter/lib/src/client_base.dart.twig [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import 'response.dart';
+import 'client.dart';
+import 'enums.dart';
+
+abstract class ClientBase implements Client {  
+{% for header in spec.global.headers %}
+{% if header.description %}
+    /// {{header.description}}
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/client_base.dart.twig:8]: Indentation 👍🏻

File: templates/flutter/lib/src/client_browser.dart.twig [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    assert(_endPoint.startsWith(RegExp("http://|https://")),
+        "endPoint $_endPoint must start with 'http'");
+    init();
+  }
+
+  String get endPoint => _endPoint;
+
+{% for header in spec.global.headers %}
+{% if header.description %}
+     /// {{header.description}}
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/client_browser.dart.twig:53]: Indentation 👍🏻

File: templates/flutter/lib/src/client_io.dart.twig [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+    final directory = await getApplicationDocumentsDirectory();
+    final path = directory.path;
+    final Directory dir = new Directory('$path/cookies');
+    await dir.create();
+    return dir;
+  }
+
+{% for header in spec.global.headers %}
+{% if header.description %}
+     /// {{header.description}}
```
Comment (TorstenDittmann) [inline templates/flutter/lib/src/client_io.dart.twig:80]: Indentation 👍🏻

**Review by TorstenDittmann — APPROVED** [review]

**Pull request merged**